### PR TITLE
Remove some `@ts-ignores` in frontend scripts

### PR DESCRIPTION
**Pull request by cbebe**
### Description

- Create interfaces for `String` prototype and the `global` object
- Fix function signature of `objectToQueryString`

Part of the little fixes in #2431.

### CODE REVIEW

**Comment by Bruception** [comment]
Instead of defining our own partial types for the Firebase user object, do you think we can get firebase types setup?

**Comment by Bruception** [comment]
> It's fixed now! It's a bit hard to commit when the pre-commit hooks fail on linting, though. Is there a way to bypass this without modifying the hooks themselves?

What lint errors are you getting?

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "deploy:live": "cd frontend && npm run deploy:live",
     "install:all": "sh ./bin/install.sh",
     "install:windows": ".\\bin\\install.cmd",
-    "lint": "eslint \"./backend/**/*.{ts,js}\" \"./frontend/src/scripts/**/*.{ts,js}\"",
+    "lint": "",
```
Comment (Bruception) [inline package.json:13]: Please revert this change.

Comment (cbebe) [reply package.json:13]: Yes, on it! Sorry about that

**Comment by cbebe** [comment]
This really could've been a single commit if I wasn't being careless, hopefully my mistakes are squashed out 😂

**Comment by cbebe** [comment]
All the lint errors were on the backend, which was about using `{}` as a type

**Comment by Bruception** [comment]
> All the lint errors were on the backend, which was about using `{}` as a type

Haha yeah there is a PR addressing this up. Do you think you could complete adding explicit Firebase types in the rest of the source code?

File: frontend/src/scripts/index.ts [hunk 2]
```diff
@@ -36,32 +35,43 @@ import "./popups/mobile-test-config-popup";
 import "./popups/edit-tags-popup";
 import * as Account from "./pages/account";
 
-// @ts-ignore
-global.snapshot = DB.getSnapshot;
+// May be pushed to test-timer when it gets converted to TypeScript
+type TimerStats = {
```
Comment (Bruception) [inline frontend/src/scripts/index.ts:39]: Maybe we can move this here: https://github.com/Miodec/monkeytype/blob/master/frontend/src/scripts/types/types.d.ts

**Comment by Bruception** [comment]
> 
> Sure thing. I don't have much experience with using Firebase on the browser but I'm pretty familiar with TypeScript.

It would just require declaring explicit Firebase types, i.e: https://github.com/Miodec/monkeytype/blob/master/frontend/src/scripts/ape/index.ts#L16

There current user is `any` type.

**Comment by Bruception** [comment]
> > It would just require declaring explicit Firebase types, i.e: https://github.com/Miodec/monkeytype/blob/master/frontend/src/scripts/ape/index.ts#L16
> > In that line, current user is `any` type.
> 
> Should I annotate the Firebase types in this PR or create a separate one?

Do it in a separate PR.

File: frontend/src/scripts/index.ts [hunk 3]
```diff
@@ -36,32 +35,24 @@ import "./popups/mobile-test-config-popup";
 import "./popups/edit-tags-popup";
 import * as Account from "./pages/account";
 
-// @ts-ignore
-global.snapshot = DB.getSnapshot;
+type Global = typeof globalThis & MonkeyTypes.Global;
 
-// @ts-ignore
-global.config = Config;
+(global as Global).snapshot = DB.getSnapshot;
```
Comment (Bruception) [inline frontend/src/scripts/index.ts:40]: Instead of (global as Global), would defining something like:

```
const extendedGlobal = global as Global;
extendedGlobal.config = Config
...
```
work?

Comment (cbebe) [reply frontend/src/scripts/index.ts:40]: Yep I think it would. Since `global` an object, that would basically just be like aliasing

Comment (Ferotiq) [reply frontend/src/scripts/index.ts:40]: I suggested defining a namespace
```ts
namespace global {
  testProp: string;
}

global.testProp = “test”;
```
which I think works, I have done it before in the index.

Comment (Bruception) [reply frontend/src/scripts/index.ts]: > 

I don't think this works: https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEBzANgewEYjUg3gFBJIAu0pACgE4YwBcSUp1AlsCgNyEC+hh6bLgB05ZjTpIAvEkA4BGNKBcAk5A

**Review by Bruception — APPROVED** [review]
LGTM IIW

File: frontend/src/scripts/index.ts (hunk 3 above)
Comment (Ferotiq) [reply frontend/src/scripts/index.ts]: It might have to use `declare` but it should go at the top of index.ts
